### ISSUE TITLE
Implement ClmFile::OpenSeekableStreamReader

### ISSUE THREAD

**Issue by Brett208** [OutpostUniverse/OP2Utility#109]
Currently, the function OpenSeekableStreamReader is only implemented on VolFile. The challenge in ClmFile is that the required header file for the music track is contained in a separate location than the body and must be read first. 

Also, some work is currently done on the header when it is extracted to create a legitimate file. So, it is not a simple read from header first, then read from body problem. (This work on the header might be avoidable if we change how the header is structured and loaded into the ClmFile, but I haven't spent any thought on it).

```std::unique_ptr<SeekableStreamReader> ClmFile::OpenSeekableStreamReader(int fileIndex)```

I would propose as a solution: 
 1. Write the entire music track into a memory buffer by constructing the header and body together.  
 2. Load the new buffer into a MemoryStreamReader
 3. Pass the MemoryStreamReader out as a unique_ptr for consumption

The code to construct the stream would parallel (maybe reuse most of) the code required when extracting a single music file. 

-Brett

**Comment by DanRStevens** [comment]
I'm not sure we want `OpenSeekableStreamReader` to produce a stream identical to a .wav file. It might make more sense to produce a stream equivalent to the contents of the data section of a .wav file. That is, a stream of PCM data. A WaveReader class might produce the same from a .wav file for going the other direction. If you want to build a .wav file from CLM output, you slap the appropriate headers onto it, and then dump the PCM stream contents.

I suppose this highlights another area for expansion of the stream library. Concatenating multiple streams together. That would be one way of building a .wav file from a header and a PCM data stream. Though, I don't think that is strictly necessary. Extract code can just manually write a header before copying the PCM stream to the output.

I think we can extract .wav data from a CLM without doing a full memory load of the PCM data, and it shouldn't be too difficult to accomplish. Especially not once we merge in code to copy data from one stream to another.

Oh, and we should probably rename `OpenSeekableStreamReader`. It's a bit wordy, and it's just naming a data type. Instead, we should describe what it does, or what the returned data represents. Perhaps `OpenPcmStream` (recommended), or `OpenWaveStream` (not recommended).

**Comment by Brett208** [comment]
Good point that OpenSeekableStreamReader is a poor name. OpenSeekableStreamReader is an inherited function from ArchiveUnpacker. We would have to destroy the virtual function to name it something music file specific as suggested. How about OpenPackedFileStream?

It is convenient to unpack a file into memory in the case of a vol archive. I don't see a use case for the functionality for a CLM file. I do not see anyone using a CLM file outside of preparing music for play in Outpost 2 or extracting it to see content. 

Perhaps what is best is to remove the virtual function from ArchiveUnpacker and only make it possible to open a stream on VolFile. Then there will be no reason to design any sort of ability to access a memory stream on a ClmFile.

I was probably mistaken in the first place to make SeekableStreamReader a part of ArchivePacker instead of just VolFile.

-Brett

**Comment by DanRStevens** [comment]
Ahh right, the name comes from the base class. What about just `OpenStream(std::string filename)`?

I think it makes sense to support this operation for both archive types, but the CLM won't return a stream that represents a complete extracted .wav file, but rather just the raw PCM data. A previous utility extracted the wave data to .raw files, which was exactly that, the PCM data, with no wave header.

I think the sticking point here, is the concept of what a stream represents. We support the operation of extracting the PCM data from a CLM file to a WAVE file. I don't think the stream coming from the CLM file needs to be in that exact format of a WAVE file. Rather, it should be the natural form used in the CLM file. It just happens that the two formats are different in this case, though similar enough to easily convert, and with .wav files in the extracted form being easier to manipulate.

**Comment by DanRStevens** [comment]
Correct, it would be just returning the "data" section, without any section headers.

Effectively, you're returning the data packed in the archive, unmodified, without conversion, without additions. It's the simplest and most flexible thing to do. It would also be consistent behaviour between VOL and CLM files. Remember that there is no WAVE file within a CLM archive. It's a different format.

Naturally though, we do want to support conversion to and from WAVE files within our tooling, but not at that level. The format conversion is minimal, easy to do, and makes working with the data much easier. However, that format conversion is independent of archive data storage operations.

**Comment by DanRStevens** [comment]
I noticed you had closed this issue while the PR was still open. I take it you meant to merge the PR.

**Comment by Brett208** [comment]
Yes, I was interrupted and hadn't gotten back to finish merging. Thank you for taking care of it.
